### DISCUSSION TITLE
Bitwise operators

### DISCUSSION THREAD

**Discussion by sumanblack666** [MilesCranmer/PySR#485] [Q&A]
Hi , I would like to know how do i use bitwise operators such as XOR, OR ,AND ? .

**Comment by MilesCranmer** [comment]
Hi @sumanblack666,

There is OR and AND already: https://astroautomata.com/PySR/operators/. 
For XOR we don't have this as a built-in but you could add it like this:

```python
    binary_operators=["logical_and", "logical_or", "logical_xor(x, y) = xor(x > 0, y > 0) ? one(x) : zero(x)"],
    extra_sympy_mappings={
        "logical_xor": lambda x, y: sympy.Piecewise((1.0, (x > 0) ^ (y > 0)), (0.0, True)),
    }
```

Cheers,
Miles

**Reply by MilesCranmer** [reply]
Edit: i just fixed the xor definition; initially it was opposite

**Reply by sumanblack666** [reply]
thank you mr.Miles .Im still running thru some errors when i run thru PySR.Basically this is what im trying to do ,i just want to prove y1,y2,y3,and y4 equation from PySR.All the values are in hex .I got these equation from DEAP , I wanted to try the same PySR

y1 = x2 ^ x3 ^ x4 ^ 0x11
y2 = x2 ^ 0x72
y3 = x3 ^ 0x80
y4 = ((x1 ^ x2 ^ x3 % 0x100) ^ x4 ^ 0x19)


<body link=blue vlink=purple>


x1 | x2 | x3 | x4 | y1 | y2 | y3 | y4
-- | -- | -- | -- | -- | -- | -- | --
63 | 7a | 1 | 42 | 28 | 8 | 81 | 7b
73 | fa | 1 | 42 | a8 | 88 | 81 | eb
83 | 26 | 1 | 42 | 74 | 54 | 81 | b7
93 | 29 | 1 | 42 | 7b | 5b | 81 | a4
a3 | a8 | 1 | 42 | fa | da | 81 | d5
b3 | 29 | 1 | 42 | 7b | 5b | 81 | 44
c3 | 75 | 1 | 42 | 27 | 7 | 81 | 20
c3 | b1 | 1 | 42 | e3 | c3 | 81 | ec
c3 | b3 | 1 | 42 | e1 | c1 | 81 | e2
d3 | 94 | 1 | 42 | c6 | e6 | 81 | f1
e3 | 82 | 1 | 42 | d0 | f0 | 81 | f3



</body>

</html>

Please help me sir .

x1 = np.array([99, 115, 131, 147, 163, 179, 195, 195, 195, 211])
x2 = np.array([122, 250, 38, 41, 168, 41, 117, 177, 179,148])
x3 = np.array([1, 1, 1, 1, 1, 1, 1, 1, 1, 1])
x4 = np.array([66, 66, 66, 66, 66, 66, 66, 66, 66,  66])
y = np.array([8, 136, 84, 91, 218, 91, 7, 195, 193, 230])


X = np.column_stack((x1, x2, x3, x4))

default_pysr_params = dict(
    populations=30,
    model_selection="best",
)

model = PySRRegressor(
    procs=2,
    binary_operators=["mod","logical_and", "logical_or", "logical_xor(x, y) = xor(x > 0, y > 0) ? one(x) : zero(x)"],
    extra_sympy_mappings={
        "logical_xor": lambda x, y: sympy.Piecewise((1.0, (x > 0) ^ (y > 0)), (0.0, True)),
    },
    #unary_operators=["neg", "square", "cube", "exp", "abs", "log", "log10", "log2", "log1p", "sqrt", "sin", "cos", "tan", "sinh","cosh","relu", "floor", "ceil", "sign" ],
    #select_k_features=5,
    complexity_of_constants=2,
    niterations=10000,
    #turbo=True
)

model.fit(X,y)

**Reply by MilesCranmer** [reply]
Oh you mean a bitwise XOR, not a boolean XOR? That would be different then.

**Reply by sumanblack666** [reply]
yes.Do you think i can use PySR to derive the equation ? or should i just stick to DEAP ?

**Reply by MilesCranmer** [reply]
You can totally do this with PySR it will just take a bit of work to define the operators that perform a bitwise XOR. Take a look at e.g., https://github.com/MilesCranmer/PySR/discussions/469 for related discussions..
